linkerd/lock/src/shared.rs

```diff
@@ -0,0 +1,166 @@
+use self::waiter::Notify;
+pub(crate) use self::waiter::Wait;
+use crate::error::Error;
+use futures::{Async, Poll};
+use std::sync::Arc;
+use tracing::trace;
+
+/// The shared state between one or more Lock instances.
+///
+/// When multiple lock instances are contending for the inner value, waiters are
+/// stored in a LIFO stack. This is done to bias for latency instead of fairness
+pub(crate) struct Shared<T> {
+    state: State<T>,
+    waiters: Vec<Notify>,
+}
+
+enum State<T> {
+    /// A Lock is holding the value.
+    Claimed,
+
+    /// The inner value is available.
+    Unclaimed(T),
+
+    /// The lock has failed.
+    Failed(Arc<Error>),
+}
+
+// === impl Shared ===
+
+impl<T> Shared<T> {
+    pub fn new(value: T) -> Self {
+        Self {
+            waiters: Vec::new(),
+            state: State::Unclaimed(value),
+        }
+    }
+
+    /// Try to claim a value without registering a waiter.
+    pub fn try_acquire(&mut self) -> Result<Option<T>, Arc<Error>> {
+        match std::mem::replace(&mut self.state, State::Claimed) {
+            // This lock has acquired the value.
+            State::Unclaimed(v) => {
+                trace!("acquired");
+                Ok(Some(v))
+            }
+            // The value is already claimed by a lock.
+            State::Claimed => Ok(None),
+            // The locks failed, so reset the state immediately so that all
+            // instances may be notified.
+            State::Failed(error) => {
+                self.state = State::Failed(error.clone());
+                Err(error)
+            }
+        }
+    }
+
+    /// Try to acquire a value or register the given waiter to be notified when
+    /// the lock is available.
+    pub fn poll_acquire(&mut self, wait: &Wait) -> Poll<T, Arc<Error>> {
+        match self.try_acquire() {
+            Ok(Some(svc)) => Ok(Async::Ready(svc)),
+            Ok(None) => {
+                // Register the current task to be notified.
+                wait.register();
+                // Register the waiter's notify handle if one isn't already registered.
+                if let Some(notify) = wait.get_notify() {
+                    trace!("Registering waiter");
+                    self.waiters.push(notify);
+                }
+                debug_assert!(wait.is_waiting());
+                Ok(Async::NotReady)
+            }
+            Err(error) => Err(error),
+        }
+    }
+
+    pub fn release_and_notify(&mut self, value: T) {
+        trace!(waiters = self.waiters.len(), "Releasing");
+        debug_assert!(match self.state {
+            State::Claimed => true,
+            _ => false,
+        });
+        self.state = State::Unclaimed(value);
+        self.notify_next_waiter();
+    }
+
+    pub fn notify_next_waiter(&mut self) {
+        while let Some(waiter) = self.waiters.pop() {
+            if waiter.notify() {
+                trace!("Notified waiter");
+                return;
+            }
+        }
```

Comment:
If I'm correctly understanding this strategy, I think we may still be potentially susceptible to some memory leakiness when a very large number of waiters are in the stack. Since this is a LIFO, if there are a lot of waiters on the lock, those towards the end of the stack may never be woken up. If the delay is long enough, those waiters will time out, causing the futures to be dropped and thus dropping the `Arc`-ed waker, but the `Weak` pointer will remain in the stack. If a lot of new waiters are always being pushed, those `Weak`s will remain at the back of the stack forever.

Since we're dropping the actual waker when a future is dropped, the _size_ of the data remaining in the stack per waiter is significantly smaller than in `tokio_sync`'s `Semaphore` (I believe a `Weak` is just a single pointer). However, it _is_ still possible for the vec to have unbounded growth.

One thought is that we may want to periodically call `Vec::retain` or something, to drop any `Notify` instances in the stack where the waiter has been dropped. However, this is expensive — it requires iterating the whole vec — so for latency reasons, we probably don't want to do this when acquiring or releasing the lock. We could probably consider putting this work in a background task that will run when the lock is not busy?

We may also want to periodically call `shrink_to_fit` so that the actual allocated capacity of the vec is released when load decreases enough that we don't need as much capacity. However, we definitely don't want to do this _too_ often, as we could essentially thrash allocations. If we're occasionally cleaning out dead waiters, vec capacity is probably much less of an issue...

Reply:
Yep. This is all correct. My current assumption is that what we have is Good Enough. We could certainly try to address these issues, but solutions all come with complexity. We can only really hit this situation when the lock is _always_ under contention. As soon as it becomes idle, the entire set of waiters will be cleaned up. It seems like it would be unlikely to encounter this in practice, though I'm sure that we could write a test that triggers this...

For now, I'd prefer to add a note about this and revisit if we see this type of usage in practice.

Reply:
> As soon as it becomes idle, the entire set of waiters will be cleaned up

Where does this happen? I must've missed it.

Reply:
Oh, you mean that if the lock is released when there are no "live" waiters in the stack, we continue popping until we hit a "live" waiter, so if there are none, we'll pop everything.  Yeah, that means we probably won't see issues around the stack's capacity except in extreme edge cases. 👍

Reply:
Added a note onto the `Shared` docstring.